### PR TITLE
SALTO-907, SALTO-921: Fix hidden values and types logic

### PR DESCRIPTION
**Pull request by ori-moisis**
- Fix cases where the hidden annotation itself changes for a type
- Fix handling of nested hidden values
- Support changes to the hidden annotation for fields (previously did not work)
- Support hidden annotation values (only for nested values)

---
This should fix some sporadic e2e failures

_Release notes_:
- Fixed issues with hidden types and values not being hidden correctly in some scenarios

### CODE REVIEW

File: packages/workspace/src/workspace/hidden_values.ts [hunk 1]
```diff
@@ -13,125 +13,327 @@
...
-    StateHiddenTypesMap
+  // If we have a change to A.B and a change to A.B.C, the second change is redundant because
+  // its value is included in the first one. a change in "unique" if there are no changes
+  // to ids that contain it
+  const hasChangeContainingId = (id: ElemID): boolean => (
+    changeById[id.getFullName()] !== undefined
+    || (!id.isTopLevel() && hasChangeContainingId(id.createParentID()))
   )
+  return Object.values(changeById)
+    .filter(change => change.id.isTopLevel() || !hasChangeContainingId(change.id.createParentID()))
```
Comment (netama) [inline packages/workspace/src/workspace/hidden_values.ts:248]: nit: do we need `change.id.isTopLevel()`? wouldn't `hasChangeContainingId(...)` be falsy?

File: packages/workspace/src/workspace/hidden_values.ts [hunk 2]
```diff
@@ -13,125 +13,327 @@
...
+): Promise<DetailedChange[]> => {
+  const filterOutHidden = async (change: DetailedChange): Promise<DetailedChange | undefined> => {
+    if (isRemovalChange(change)) {
+      // There should be no harm in letting remove changes through here. remove should be resilient
+      // to its subject not existing
+      return change
+    }
+
+    const { parent, path } = change.id.createTopLevelParentID()
+    const baseElem = change.id.isTopLevel() ? change.data.after : await state.get(parent)
```
Comment (netama) [inline packages/workspace/src/workspace/hidden_values.ts:284]: can you please explain the `isTopLevel` case?

File: packages/workspace/src/workspace/hidden_values.ts [hunk 3]
```diff
@@ -13,125 +13,327 @@
...
+
+    if (baseElem === undefined) {
+      // If something is not in the state it cannot be hidden
```
Comment (netama) [inline packages/workspace/src/workspace/hidden_values.ts:287]: when would this happen?

File: packages/adapter-utils/src/compare.ts [hunk 4]
```diff
@@ -154,3 +155,13 @@ export const detailedCompare = (
     : []
   return [...annotationTypeChanges, ...annotationChanges, ...fieldChanges]
 }
+
+export const applyDetailedChanges = (
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/adapter-utils/src/compare.ts:159]: felt like it is coupled with the `detailedCompare` function that is the main function exported from this file and I didn't want to just put it in "utils.ts" like everything else.

any suggestions for how to structure it?
maybe rename this file to `detailed_change.ts` or `change.ts` because it has functions that deal with detailed changes?

File: packages/workspace/src/workspace/hidden_values.ts [hunk 5]
```diff
@@ -13,125 +13,327 @@
...
+  // When a field is made visible, all the values of that field need to be added
+  // In order to support making a field visible we must traverse all state elements
+  // to find all the hidden values we need to add.
+  // Theoretically in order to hide values we would need to iterate the workspace elements
+  // but since we assume a redundant remove change is handled gracefully, we optimize this
+  // by only iterating the state elements and creating remove changes for values that are
+  // newly hidden, some of these remove changes may be redundant (if the value we
+  // are trying to hide was already removed manually from the workspace element)
+  const workspaceElementIds = new Set(
+    (await getWorkspaceElements()).map(element => element.elemID.getFullName())
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:216]: I am fairly sure that in all cases where this will actually be called (only inside `updateNacls`) this would not be the first time `getAll` is called and therefore all the elements would already be loaded into memory (they are cached after the first call to `getAll()`) so this would just be returning a pointer.

File: packages/workspace/src/workspace/hidden_values.ts [hunk 6]
```diff
@@ -13,125 +13,327 @@
...
+
+  const isPartOfHiddenPath = (path?: ElemID): boolean => (
+    // Something is considered a part of a hidden path if it is a prefix of a hidden path
+    // or if it is nested inside a hidden path.
+    // Assume A.B.C is a hidden path, when we transform the element we must not omit A.B
+    // because then we'd never reach A.B.C
+    // We also do not omit A.B.C.D because A.B.C was hidden, so by association everything inside
+    // it is also hidden
+    path !== undefined && hiddenPaths.some(hiddenPath => (
+      hiddenPath.startsWith(path.getFullName()) || path.getFullName().startsWith(hiddenPath)
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:79]: good point.
I'll make the fix more semantic so that instead of relying on the format of `getFullName` it will use the actual element ID to check if something is nested inside something else

File: packages/workspace/src/workspace/hidden_values.ts [hunk 7]
```diff
@@ -13,125 +13,327 @@
...
+): Values | undefined => (
+  transformValues({
+    values: value,
+    type,
+    transformFunc: removeHiddenValue,
+    pathID,
+  })
+)
+
+const isChangeToHidden = (change: DetailedChange): boolean => (
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:127]: the problem is that this logic is correct iff the "default" of the thing being checked is to be `false` if it doesn't exist, calling it `isChangeToTrue` (and the other one would be `isChangeToFalse` I guess) would be misleading (specifically `isChangeToFalse` won't actually check that something is `false`).

I'll move the `change.id.name` condition here (it will cause it to be checked twice, but that is not a big deal...

File: packages/workspace/src/workspace/hidden_values.ts (hunk 1 above)
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:248]: technically it is possible to call `createParentID` on a top level ID (currently it would return the adapter's ID).
while I can't think of a scenario where we would have a change with an adapter's ID, it felt safer to avoid checking for it in the first place.
also I feel that the definition for what should be the answer to `createParentID` on a top level ID is a little fuzzy (should it be the adapter? should it just be the same ID because top level IDs have no parent?) and I didn't want to rely on the current implementation if I didn't have to.

I will try to refactor the code a little more and hopefully get to something more clear.
if it is still unclear, another option would be to change this code to be more readable but significantly less efficient:
```typescript
  const hasChangeOnParent = (id: ElemID): boolean => (
    Object.values(changeById).some(change => change.id.isParentOf(id))
  )
  return Object.values(changeById)
    .filter(change => !hasChangeOnParent(change.id))
```

File: packages/workspace/src/workspace/hidden_values.ts (hunk 2 above)
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:284]: small optimization - if the change is on a top level element (as would be the case with a completely new hidden type for example), the change data already holds the top level element so we don't have to go to the state in order to get it.

File: packages/workspace/src/workspace/hidden_values.ts (hunk 3 above)
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:287]: can't really think of a case where this would happen (except for the tests where we sometimes mock an incomplete workspace).

File: packages/workspace/src/workspace/hidden_values.ts [hunk 8]
```diff
@@ -13,125 +13,327 @@
...
+  const changes = [...origChanges, ...additionalChanges]
+  // If we have multiple changes on the same ID, we want to take the newer change
+  // (the one from additionalChanges) because it overrides the original, for example
+  // this can happen if a field is both hidden and changed at the same time
+  const changeById = _.keyBy(changes, change => change.id.getFullName())
 
-  const elementsWithUpdatedTypes = updateMergedTypes(
-    instancesWithHiddenValues,
-    StateHiddenTypesMap
+  // If we have a change to A.B and a change to A.B.C, the second change is redundant because
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:247]: avoid what?

File: packages/workspace/src/workspace/hidden_values.ts [hunk 9]
```diff
@@ -13,125 +13,336 @@
...
+    }
+
+    if (isInstanceElement(baseElem) || (isObjectType(baseElem) && change.id.idType === 'attr')) {
+      // Instance values and annotation values can be hidden
+      const [changeType, valuePath] = isInstanceElement(baseElem)
+        ? [baseElem.type, path]
+        : [baseElem.annotationTypes[path[0]], path.slice(1)]
+
+      if (changeType === undefined) {
+        // a value without a type cannot be hidden
```
Comment (netama) [inline packages/workspace/src/workspace/hidden_values.ts:312]: nit: is this a valid scenario? if not - might be worth logging (debug)?

File: packages/workspace/src/workspace/hidden_values.ts [hunk 10]
```diff
@@ -13,125 +13,336 @@
...
+      hiddenValueChanges.push(createAddChange(value, path))
+      return undefined
+    }
+    return value
+  }
 
-  const StateHiddenTypesMap = _.keyBy(
-    stateElements.filter(isHiddenType) as TypeElement[],
-    o => o.elemID.getFullName()
+  // When a field becomes hidden, all the values of that field need to be removed
```
Comment (netama) [inline packages/workspace/src/workspace/hidden_values.ts:213]: nit: I think if the first paragraph here was a docstring for the top-level function, it would make the high-level flow a little easier to understand

**Review by netama — APPROVED** [review]
Logic looks good (I might do another pass on `filterOutHiddenChanges` later but I'm sure that was covered).
I think `hidden_values.ts` has complex logic that is a little hard to follow without knowing all the details up front, added some nit suggestions that would make it easier for me personally but completely up to you.

File: packages/workspace/src/workspace/hidden_values.ts (hunk 2 above)
Comment (netama) [reply packages/workspace/src/workspace/hidden_values.ts:293]: no objections to keeping it, just wanted to make sure I didn't miss anything :)

File: packages/workspace/src/workspace/hidden_values.ts (hunk 9 above)
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:312]: it is valid - can happen for example if we have an annotation value that doesn't have a defined type (for example the built in annotation `_default` never has a type).

File: packages/workspace/src/workspace/hidden_values.ts [hunk 11]
```diff
@@ -13,125 +13,336 @@
...
+  }
 
-  return elementsWithUpdatedTypes.concat(hiddenTypes)
+  return Object.values(changeById).filter(change => !hasChangeOnParent(change.id))
 }
 
-export const removeHiddenFieldValue: TransformFunc = ({ value, field }) => (
-  isHidden(field) ? undefined : value
-)
+const handleChangesToHiddenAnnotation = async (
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:260]: the reason I called it "handle" (also couldn't think of a clearer name) is because this can also filter out some of the original changes (meaning, it doesn't just add new changes).

maybe "mergeWithHiddenChangesSideEffects"?

File: packages/workspace/src/workspace/hidden_values.ts [hunk 12]
```diff
@@ -13,125 +13,336 @@
...
+export const handleHiddenChanges = async (
+  changes: DetailedChange[],
+  state: State,
+  getWorkspaceElements: () => Promise<Element[]>,
+): Promise<DetailedChange[]> => {
+  const changesWithHiddenAnnotationChanges = await handleChangesToHiddenAnnotation(
+    changes, state, getWorkspaceElements,
+  )
+  return filterOutHiddenChanges(changesWithHiddenAnnotationChanges, state)
+}
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/workspace/src/workspace/hidden_values.ts:348]: the file currently has only 2 exported functions: `mergeWithHidden` which handles one direction (merging nacl with hidden stuff) and `handleHiddenChanges` which handles the other direction (splitting changes to nacl from changes to hidden stuff).
they are pretty much completely separate in terms of implementation (they share almost no code) and in the file (all the "helper functions" for `mergeWithHidden` are above it, everything below is for `handleHiddenChanges`).
just from a style perspective I don't like to create separation between the function and it's helper functions, so I don't want to structure the file like this
```typescript
// mergeWithHidden helper functions
// handleHiddenChanges helper functions

export const mergeWithHidden = ...
export const handleHiddenChanges = ...
```
because then navigating the `mergeWithHidden` implementation would be harder than it has to be (would have to jump farther in the file to go back and forth between the main function and its implementation details).

I could split off `mergeWithHidden` to a different file, but its implementation is so short that I don't think there will be much value in that...
